File: eHour-persistence/src/main/java/net/rrm/ehour/persistence/project/dao/ProjectAssignmentDaoHibernateImpl.java
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

package net.rrm.ehour.persistence.project.dao;

import com.google.common.base.Optional;
import net.rrm.ehour.data.DateRange;
import net.rrm.ehour.domain.*;
import net.rrm.ehour.persistence.dao.AbstractGenericDaoHibernateScalaImpl;
import org.hibernate.Criteria;
import org.hibernate.criterion.Restrictions;
import org.springframework.stereotype.Repository;

import java.util.List;

/**
 * CRUD stuff on PA do
 */
@Repository("projectAssignmentDao")
public class ProjectAssignmentDaoHibernateImpl
        extends AbstractGenericDaoHibernateScalaImpl<ProjectAssignment, Integer>
        implements ProjectAssignmentDao {
    private static final Optional<String> CACHEREGION = Optional.of("query.ProjectAssignment");

    public ProjectAssignmentDaoHibernateImpl() {
        super(ProjectAssignment.class);
    }

    @Override
    public List<ProjectAssignment> findProjectAssignmentsForUser(Integer userId, DateRange range) {
        List<ProjectAssignment> results;

        String[] keys = new String[]{"dateStart", "dateEnd", "userId"};
        Object[] params = new Object[]{range.getDateStart(), range.getDateEnd(), userId};

        results = findByNamedQueryAndNamedParam("ProjectAssignment.findProjectAssignmentsForUserInRange"
                , keys
                , params
                , CACHEREGION);


        return results;
    }

    @Override
    public List<ProjectAssignment> findProjectAssignmentForUser(Integer projectId, Integer userId) {
        String[] keys = new String[]{"projectId", "userId"};
        Integer[] params = new Integer[]{projectId, userId};
        List<ProjectAssignment> results;

        results = findByNamedQueryAndNamedParam("ProjectAssignment.findProjectAssignmentsForUserForProject"
                , keys
                , params
                , CACHEREGION);

        return results;
    }

    @Override
    public List<ProjectAssignment> findProjectAssignmentsForUser(User user) {
        List<ProjectAssignment> results;

        results = findByNamedQueryAndNamedParam("ProjectAssignment.findProjectAssignmentsForUser"
                , "user"
                , user
                , CACHEREGION);

        return results;
    }

    @Override
    public List<ProjectAssignment> findProjectAssignmentsForProject(Project project, DateRange range) {
        List<ProjectAssignment> results;
        String[] keys = new String[]{"dateStart", "dateEnd", "project"};
        Object[] params = new Object[]{range.getDateStart(), range.getDateEnd(), project};

        results = findByNamedQueryAndNamedParam("ProjectAssignment.findProjectAssignmentsForProjectInRange"
                , keys
                , params
                , CACHEREGION);

        return results;
    }

    @Override
    public List<ProjectAssignment> findAllProjectAssignmentsForProject(Project project) {
        return findProjectAssignmentsForProject(project, false);
    }

    @Override
    public List<ProjectAssignment> findAllActiveProjectAssignmentsForProject(Project project) {
        return findProjectAssignmentsForProject(project, true);
    }

    @SuppressWarnings("unchecked")
    private List<ProjectAssignment> findProjectAssignmentsForProject(Project project, Boolean onlyActive) {
        Criteria crit = getSession().createCriteria(ProjectAssignment.class);

        if (onlyActive) {
            crit.add(Restrictions.eq("active", true));
        }

        crit.add(Restrictions.eq("project", project));

        return crit.list();
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<ProjectAssignmentType> findProjectAssignmentTypes() {
        return (List<ProjectAssignmentType>)getSession().createCriteria(ProjectAssignmentType.class).list();
    }

    @Override
    public List<ProjectAssignment> findProjectAssignmentsForCustomer(Customer customer, DateRange range) {
        List<ProjectAssignment> results;
        String[] keys = new String[]{"dateStart", "dateEnd", "customer"};
        Object[] params = new Object[]{range.getDateStart(), range.getDateEnd(), customer};

        results = findByNamedQueryAndNamedParam("ProjectAssignment.findProjectAssignmentsForCustomerInRange"
                , keys
                , params
                , CACHEREGION);

        return results;
    }
}

File: eHour-persistence/src/main/scala/net/rrm/ehour/persistence/project/dao/ProjectAssignmentDaoHibernateImpl.scala
package net.rrm.ehour.persistence.project.dao

import java.util

import net.rrm.ehour.data.DateRange
import net.rrm.ehour.domain._
import net.rrm.ehour.persistence.dao.AbstractGenericDaoHibernateImpl
import net.rrm.ehour.persistence.retry.ExponentialBackoffRetryPolicy
import org.hibernate.criterion.Restrictions
import org.springframework.stereotype.Repository

/**
 * CRUD stuff on PA do
 */
@Repository("projectAssignmentDao")
class ProjectAssignmentDaoHibernateImpl extends AbstractGenericDaoHibernateImpl[Integer, ProjectAssignment](classOf[ProjectAssignment]) with ProjectAssignmentDao {
  private final val CacheRegion = Some("query.ProjectAssignment")

  override def findProjectAssignmentsForUser(userId: Integer, range: DateRange): util.List[ProjectAssignment] = {
    val keys = List("dateStart", "dateEnd", "userId")
    val params = List(range.getDateStart, range.getDateEnd, userId)
    findByNamedQuery("ProjectAssignment.findProjectAssignmentsForUserInRange", keys, params, CacheRegion)
  }

  override def findProjectAssignmentForUser(projectId: Integer, userId: Integer): util.List[ProjectAssignment] = {
    val keys = List("projectId", "userId")
    val params  = List(projectId, userId)
    findByNamedQuery("ProjectAssignment.findProjectAssignmentsForUserForProject", keys, params, CacheRegion)
  }

  override def findProjectAssignmentsForUser(user: User): util.List[ProjectAssignment] =
    findByNamedQuery("ProjectAssignment.findProjectAssignmentsForUser", "user", user, CacheRegion)

  override def findProjectAssignmentsForProject(project: Project, range: DateRange): util.List[ProjectAssignment] = {
    val keys = List("dateStart", "dateEnd", "project")
    val params = List(range.getDateStart, range.getDateEnd, project)
    findByNamedQuery("ProjectAssignment.findProjectAssignmentsForProjectInRange", keys, params, CacheRegion)
  }

  override def findAllProjectAssignmentsForProject(project: Project): util.List[ProjectAssignment] = findProjectAssignmentsForProject(project, onlyActive = false)

  override def findAllActiveProjectAssignmentsForProject(project: Project): util.List[ProjectAssignment] = findProjectAssignmentsForProject(project, onlyActive = true)

  private def findProjectAssignmentsForProject(project: Project, onlyActive: Boolean): util.List[ProjectAssignment] = {
    val crit = getSession.createCriteria(classOf[ProjectAssignment])

    if (onlyActive) {
      crit.add(Restrictions.eq("active", true))
    }

    crit.add(Restrictions.eq("project", project))

    ExponentialBackoffRetryPolicy retry crit.list.asInstanceOf[util.List[ProjectAssignment]]
  }

  override def findProjectAssignmentTypes(): util.List[ProjectAssignmentType] =
    ExponentialBackoffRetryPolicy retry getSession.createCriteria(classOf[ProjectAssignmentType]).list.asInstanceOf[util.List[ProjectAssignmentType]]

  override def findProjectAssignmentsForCustomer(customer: Customer, range: DateRange): util.List[ProjectAssignment] = {
    val keys = List("dateStart", "dateEnd", "customer")
    val params = List(range.getDateStart, range.getDateEnd, customer)
    findByNamedQuery("ProjectAssignment.findProjectAssignmentsForCustomerInRange", keys, params, CacheRegion)
  }
}
